kernel/Utilities/ElementLocalIndexing.h

```diff
@@ -0,0 +1,152 @@
+}
+
+namespace Utilities {
+
+/**
+ * Index of the local DoFs on an Element.
+ *
+ * For element matrices and vectors the rows and columns correspond to DoFs of
+ * a particular (sub)set of unknowns. Which are the DoFs for these unknowns that
+ * have support on the element. This class provides easy access to information
+ * about the local ordering of the DoFs in matrices and vectors.
+ */
+class ElementLocalIndexing {
+   public:
+    /// Create a ElementLocalIndexing without any includedUnknowns.
+    /// \param numberOfUnknowns The largest unknown that can be used.
+    explicit ElementLocalIndexing(std::size_t numberOfUnknowns);
+    /// Update the included unknowns. The unknowns should be unique and less
+    /// than the maximum.
+    ///
+    /// \param includedUnknowns The new set of included unknowns.
+    void reinit(const std::vector<std::size_t>& includedUnknowns);
+    /// Reinit the values for a different element
+    /// \param element The new element (may be nullptr)
+    void reinit(const Base::Element* element);
+
+    /// Combined reinit for both the element and included unknowns.
+    ///
+    /// \param element The new element (may be nullptr)
+    /// \param includedUnknowns The new set of included unknowns.
+    void reinit(const Base::Element* element,
+                const std::vector<std::size_t>& includedUnknowns) {
+        reinit(includedUnknowns);
+        reinit(element);
+    }
+
+    /// Get the Offset in an ElementMatrix/Vector for the unknown. Will be an
+    /// invalid value for non included unknowns or when no element is available.
+    std::size_t getDoFOffset(std::size_t unknown) const {
+        logger.assert_debug(unknown < numberOfUnknowns_,
+                            "Unknown % larger than maximum %", unknown,
+                            numberOfUnknowns_);
+        return offsets_[unknown];
+    }
+    /// Get the number DoFs (rows/columns) for the unknown in an element
+    /// matrix/vector. Will be zero for non included unknowns or when no element
+    /// is available.
+    std::size_t getNumberOfDoFs(std::size_t unknown) const {
+        logger.assert_debug(unknown < numberOfUnknowns_,
+                            "Unknown % larger than maximum %", unknown,
+                            numberOfUnknowns_);
+        return sizes_[unknown];
+    }
+    /// Total number of DoFs of the included unknowns and thus rows/columns of
+    /// an element matrix or vector. Without element this will be zero.
+    std::size_t getNumberOfDoFs() const { return totalNumberOfDofs_; }
+
+    /// Get the current element.
+    const Base::Element* getElement() const { return element_; }
+    /// The unknowns that are included.
+    const std::vector<std::size_t>& getIncludedUnknowns() const {
+        return includedUnknowns_;
+    }
+
+    /// Check the invariants of the datastructure to see if the internal
+    /// constraints are satisfied. To be used for testing and debugging.
+    void validate() const;
```

Comment:
I am not a big fan of exposting API functions that are only used for testing.

Reply:
For me it depends. I have moved most functionality out of this method to the actual test. But there are several cases where (especially in concurent/MPI settings), where in debugging cases it is easy to test whether the state is valid. Moreover, this way ensures that the internal invariants are stated somewhere.

Reply:
okay